Issue 100: Add a settings menu with a single preference for controlling sync frequency.

jwir3 opened the pull request:
I added a new 'Settings' menu to the Dashboard Activity which contains a single preference: the frequency, in minutes, of how often to synchronize repositories.

This works, but it's sort of a first pass at fixing this problem. Ideally, I think it would be nice to be able to sync with different options (e.g. every X minutes/hours, once per week, once a day, X day of the month, etc..). This is possible right now, but it requires that the user calculate the sync frequency to minutes, which is rather tedious for longer-term syncs.

rtyley commented on .gitignore:
```diff
@@ -25,3 +25,5 @@ out
 agit/artwork/rasters/
 agit/gravatars
+
+agit/lint.xml
```
I don't think this needs to be part of this commit? Feel free to add a lint.xml in a separate PR if it's useful.

jwir3 replied:
Sure, I'll remove this. I think it snuck in when I was importing it to Android Studio.

jwir3 committed:
Issue #100: Add a settings screen with a single setting that controls sync frequency.

rtyley commented:
Thanks for this submission - it's a good feature, but I think this dialog would be better as an exponential slider, or just a combo box with decent options (which I think is what you mean by this being a first pass):

![screenshot_2014-03-02-12-29-49](https://f.cloud.github.com/assets/52038/2305509/4e86873a-a25a-11e3-8e69-55f4ac5f5af2.png)

I don't think users really _need_ to be able to set their sync time to minute precision- a reasonable set of effective options would be:
- every minute
- every 5 minutes
- every 15 minutes
- every hour
- every 4 hours
- every day
- every week

...and I can't imagine a situation where it would benefit the user to have longer sync periods. 

Incidentally, the CLA for Agit is here, if you want to make a submission please make sure you specify which copyright option you choose:

https://github.com/rtyley/agit/wiki/Contribution-Agreement

jwir3 commented:
I've updated the pull request to remove the lint.xml from the .gitignore file. Additionally, I caught a problem with the AndroidManifest.xml file where I inadvertently changed the versionCode field, which I've now reverted. 

I'll be happy to make the change outlined in your most recent comment, Roberto. 

As for the question on whether I would assign the copyright to you or disclaim the copyright, and thus put the code in the public domain, is there really a difference? My suspicion is that by assigning the copyright to you, it would add the ability for you to go after someone for outright copyright infringement if the need arose, whereas the other option would not have this available. Is this correct? (There might be something else I'm missing, so I just wanted to make sure I understand the two options fully).

jwir3 committed:
Issue #100: Added Scott Johnson to CREDITS.markdown.

rtyley commented:
> I'll be happy to make the change outlined in your most recent comment, Roberto.

That's great :sparkles: For instance [PocketCast](https://play.google.com/store/apps/details?id=au.com.shiftyjelly.pocketcasts&hl=en_GB) does this, totally reasonably:

![screenshot_2014-03-02-23-31-12 1](https://f.cloud.github.com/assets/52038/2305657/ed250954-a262-11e3-88d1-07b868f6e5c2.png)

> As for the question on whether I would assign the copyright to you or disclaim the copyright, and thus put the code in the public domain, is there really a difference?

Yeah, the reasoning is covered pretty fully here:

http://www.gnu.org/licenses/gpl-faq.html#AssignCopyright

jwir3 committed:
Issue #100: Modified sync frequency preference screen to open a new dialog with different selections of sync options.

jwir3 committed:
Issue #100: Added logic to support time-of-day periodic synchronization with git repositories.

jwir3 committed:
Issue #100: Implemented time-of-day option for syncing by creating an alarm that will run a sync immediately when the alarm goes off.

jwir3 commented:
I think this is ready to be pulled. I've mimicked the example screen, and enabled a version of sync that happens at time-of-day using the `AlarmManager`.

Also, I hereby transfer copyright to Roberto Tyley for this feature to be included in agit.

rtyley commented on CREDITS.markdown:
```diff
@@ -43,6 +43,8 @@ the alternative [Git logo](http://henrik.nyh.se/2007/06/alternative-git-logo-and
 
 *   Bernd Hirschler - [German Translation](https://github.com/rtyley/agit/pull/35)
 
+*   Scott Johnson - [Sync Frequency UI](https://github.com/rtyler/agit/pull/103)
```
It's @rtyley - not rtyler!

jwir3 replied:
Oh, I'm very sorry. It was a typo... I'll change it as soon as I have a
chance.

~Scott
On Mar 15, 2014 3:56 PM, "Roberto Tyley" notifications@github.com wrote:

> In CREDITS.markdown:
> 
> > @@ -43,6 +43,8 @@ the alternative [Git logo](http://henrik.nyh.se/2007/06/alternative-git-logo-and
> > -   Bernd Hirschler - [German Translation](https://github.com/rtyley/agit/pull/35)
> > 
> > +\*   Scott Johnson - [Sync Frequency UI](https://github.com/rtyler/agit/pull/103)
> 
> It's @rtyley https://github.com/rtyley - not rtyler!
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/rtyley/agit/pull/103/files#r10636607
> .

rtyley commented on agit/res/values/strings.xml:
```diff
@@ -112,4 +113,38 @@
+    <string name="setting_sync_frequency_daily_min_key">setting_sync_frequency_daily_min</string>
+
+    <string-array name="setting_sync_choices">
+        <item>Manually</item>
```
I should apologise here, as I should have specified in my comment https://github.com/rtyley/agit/pull/103#issuecomment-36471876 that while I thought the screenshot represented a reasonable look-and-feel (ie using a restricted set of options in a list box), the preferred list of options was actually mentioned in https://github.com/rtyley/agit/pull/103#issuecomment-36471104, ie the list would be this:
- Manually
- Schedule daily at... (I'm in two minds about this, but on balance, fair enough)
- Every minute
- Every 5 minutes
- Every 15 minutes
- Every hour
- Every 4 hours
- Every day
- Every week

I wasn't actually anticipating you implementing the 'Schedule daily at...' functionality, and initially I was going to ask you to take it out, because I wasn't sure it was worth the extra code and requirement for an additional permission. I'm still rather ambivalent about it, but if you think it's worth having, then it can stay in...

jwir3 replied:
I've now updated the list of sync possibilities to include the options (and only those options) that you specify in this list.

> I wasn't actually anticipating you implementing the 'Schedule daily at...' functionality, and initially I was going to ask you to take it out, because I wasn't sure it was worth the extra code and requirement for an additional permission. I'm still rather ambivalent about it, but if you think it's worth having, then it can stay in...

Oh, my mistake. I didn't realize you weren't thinking this would go into the change. I actually think this is the one I'll use most frequently. I use agit quite frequently for updating my configuration repository. Since it's on my phone, I can then just pull updated configuration files off of my phone onto another machine when doing sysadmin tasks (e.g. if it doesn't have internet access for some reason, I can use my phone as a USB drive quite easily).

Since I do update these configuration files remotely, but not very often, I thought it would be nice if it sync'ed every night. However, since there are a lot of them, I'd like to sync every night when I'm sleeping (and thus not using my phone). :)

On the other hand, I can totally see that the added permission is not ideal. It's up to you - I'd be fine taking it out if you would rather not incur the cost of the app permission.

rtyley commented on agit/res/layout/settings_activity.xml:
```diff
@@ -0,0 +1,18 @@
+            android:id="@+id/setting_sync_frequency_listPref"
+            android:key="@string/setting_sync_frequency_key"
+            android:title="@string/setting_title_sync_frequency"
+            android:summary="@string/setting_instruction_sync_frequency"
```
The summary text shown to the user on the 'Settings' screen should actually tell them what the schedule is, so they don't have to tap through to see what it is. Ie 'Every Hour' (PocketCast again):

![screenshot_2014-03-15-22-36-07](https://f.cloud.github.com/assets/52038/2429695/6ce2da84-ac92-11e3-82e3-429e68829967.png)

jwir3 replied:
I've updated this now to perform this action. It's included in the most recent diff.

rtyley commented on agit/res/values/strings.xml:
```diff
@@ -112,4 +113,38 @@
+        <item>Every 3 hours</item>
+        <item>Every 6 hours</item>
+        <item>Every 12 hours</item>
+        <!-- <item>Every week on&#8230;</item> -->
```
Please remove these unnecessary items (the commented out options) - from a user experience perspective, I don't think they make any sense (grab Git repos every Monday? what's wrong with daily or weekly?), and you should always prefer to remove unnecessary source code.

jwir3 replied:
Ah yes I meant to remove these. My mistake. I'll pull them out in the next
revision.

~Scott
On Mar 15, 2014 5:43 PM, "Roberto Tyley" notifications@github.com wrote:

> In agit/res/values/strings.xml:
> 
> > -    <string name="setting_sync_frequency_key">setting_sync_frequency</string>
> >   +
> > -    <string name="setting_sync_frequency_daily">0</string>
> > -    <string name="setting_sync_frequency_daily_title">Sync daily at</string>
> > -    <string name="setting_sync_frequency_daily_hour_key">setting_sync_frequency_daily_hour</string>
> > -    <string name="setting_sync_frequency_daily_min_key">setting_sync_frequency_daily_min</string>
> >   +
> > -    <string-array name="setting_sync_choices">
> > -        <item>Manually</item>
> > -        <item>Schedule daily at&#8230;</item>
> > -        <item>Every 15 minutes</item>
> > -        <item>Every hour</item>
> > -        <item>Every 3 hours</item>
> > -        <item>Every 6 hours</item>
> > -        <item>Every 12 hours</item>
> > -        <!-- <item>Every week on&#8230;</item> -->
> 
> Please remove these unnecessary items (the commented out options) - from a
> user experience perspective, I don't think they make any sense (grab Git
> repos every Monday? what's wrong with daily or weekly?), and you should
> always prefer to remove unnecessary source code.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/rtyley/agit/pull/103/files#r10637045
> .

jwir3 committed:
Issue #100: Made a few changes that were found during review of PR #103.

The following items were changed:
  * Removed commented options to sync on a specific day of the week and sync on
    a specific day of the month.
  * Added additional possibilities to sync every: 5 minutes, 4 hours, daily (not
    at a specified time), and weekly.
  * Added some summary text to the list preference in the SettingsActivity that
    displays the current sync setting, so the user doesn't have to click into
    the sub-screen to see which sync setting is selected.

jwir3 commented:
One thing I've noticed when testing this is that on repos that have a self signed cert, sync can't ask for permission. I'm not sure if this should be a separate ticket, but this hampers sync in these cases because there isn't a way to prevent agit from asking if the cert is OK. Perhaps there should be a cert storage mechanism, or something cooked into this ticket that allows users to just accept sync operations on a given repository once the first manual refresh has been done?

jwir3 commented:
@rtyley Aside from issue #107, this pull request is probably ready for review again. I have confirmed that on hosts that don't have the issue described in #107 (i.e. hosts that don't have a self-signed certificate), the sync options all work as expected.

jwir3 commented:
Review ping?
